// knx-dpt/src/layouts/dpt14.js
module.exports = {
    id: 14,
    base: {
        desc: "32-bit floating point value",
        beforeDeserialize: 0,
        props: [
            {type: "float", size: 32, index: ""}
        ]
    },
    subs: {
        "000": {name: "DPT_Value_Acceleration", unit: "ms-2 1 ms-2 acceleration", use: "G"},
        "001": {name: "DPT_Value_Acceleration_Angular", unit: "rad s-2",desc: "acceleration, angular", use: "G"},
        "002": {name: "DPT_Value_Activation_Energy", unit: "J mol-1",desc: "mol-1 activation energy", use: "G"},
        "003": {name: "DPT_Value_Activity", unit: "s-1",desc: "activity (radioactive)", use: "G"},
        "004": {name: "DPT_Value_Mol", unit: "mol",desc: "amount of substance", use: "G"},
        "005": {name: "DPT_Value_Amplitude", unit: "- - amplitude (unit as appropriate)", use: "G"},
        "006": {name: "DPT_Value_AngleRad", unit: "rad",desc: "angle, radiant", use: "G"},
        "007": {name: "DPT_Value_AngleDeg", unit: "° 1 ° angle, degree", use: "G"},
        "008": {name: "DPT_Value_Angular_Momentum", unit: "J s",desc: "angular momentum", use: "G"},
        "009": {name: "DPT_Value_Angular_Velocity", unit: "rad s-1",desc: "angular velocity", use: "G"},
        "010": {name: "DPT_Value_Area", unit: "m2",desc: "area", use: "G"},
        "011": {name: "DPT_Value_Capacitance", unit: "F",desc: "capacitance", use: "G"},
        "012": {name: "DPT_Value_Charge_DensitySurface", unit: "C m-2",desc: "charge density (surface)", use: "G"},
        "013": {name: "DPT_Value_Charge_DensityVolume", unit: "C m-3",desc: "charge density (volume)", use: "G"},
        "014": {name: "DPT_Value_Compressibility", unit: "m2 N-1",desc: "compressibility", use: "G"},
        "015": {name: "DPT_Value_Conductance", unit: "S = Ω-1",desc: "conductance", use: "G"},
        "016": {name: "DPT_Value_Electrical_Conductivity", unit: "S m-1",desc: "conductivity, electrical", use: "G"},
        "017": {name: "DPT_Value_Density", unit: "kg m-3",desc: "density", use: "G"},
        "018": {name: "DPT_Value_Electric_Charge", unit: "C",desc: "electric charge", use: "G"},
        "019": {name: "DPT_Value_Electric_Current", unit: "A",desc: "electric current", use: "G"},
        "020": {name: "DPT_Value_Electric_CurrentDensity", unit: "A m-2",desc: "electric current density", use: "G"},
        "021": {name: "DPT_Value_Electric_DipoleMoment", unit: "C m",desc: "electric dipole moment", use: "G"},
        "022": {name: "DPT_Value_Electric_Displacement", unit: "C m-2",desc: "electric displacement", use: "G"},
        "023": {name: "DPT_Value_Electric_FieldStrength", unit: "V m-1",desc: "electric field strength", use: "G"},
        "024": {name: "DPT_Value_Electric_Flux", unit: "c",desc: "electric flux", use: "G"},
        "025": {name: "DPT_Value_Electric_FluxDensity", unit: "C m-2",desc: "electric flux density", use: "G"},
        "026": {name: "DPT_Value_Electric_Polarization", unit: "C m-2",desc: "electric polarization", use: "G"},
        "027": {name: "DPT_Value_Electric_Potential", unit: "V",desc: "electric potential", use: "G"},
        "028": {name: "DPT_Value_Electric_PotentialDifference", unit: "V",desc: "electric potential difference", use: "G"},
        "029": {name: "DPT_Value_ElectromagneticMoment", unit: "A m2",desc: "electromagnetic moment", use: "G"},
        "030": {name: "DPT_Value_Electromotive_Force", unit: "V",desc: "electromotive force", use: "G"},
        "031": {name: "DPT_Value_Energy", unit: "J",desc: "energy", use: "G"},
        "032": {name: "DPT_Value_Force", unit: "N",desc: "force", use: "G"},
        "033": {name: "DPT_Value_Frequency", unit: "Hz = s-1",desc: "frequency", use: "G"},
        "034": {name: "DPT_Value_Angular_Frequency", unit: "rad s-1",desc: "frequency, angular (pulsatance)", use: "G"},
        "035": {name: "DPT_Value_Heat_Capacity", unit: "J K-1",desc: "heat capacity", use: "G"},
        "036": {name: "DPT_Value_Heat_FlowRate", unit: "W",desc: "heat flow rate", use: "G"},
        "037": {name: "DPT_Value_Heat_Quantity", unit: "J",desc: "heat, quantity of", use: "G"},
        "038": {name: "DPT_Value_Impedance", unit: "Ω 1 Ω impedance", use: "G"},
        "039": {name: "DPT_Value_Length", unit: "m",desc: "length", use: "G"},
        "040": {name: "DPT_Value_Light_Quantity", unit: "J or lm s",desc: "light, quantity of", use: "G"},
        "041": {name: "DPT_Value_Luminance", unit: "cd m-2",desc: "luminance", use: "G"},
        "042": {name: "DPT_Value_Luminous_Flux", unit: "lm",desc: "luminous flux", use: "G"},
        "043": {name: "DPT_Value_Luminous_Intensity", unit: "cd",desc: "luminous intensity", use: "G"},
        "044": {name: "DPT_Value_Magnetic_FieldStrength", unit: "A m-1",desc: "magnetic field strength", use: "G"},
        "045": {name: "DPT_Value_Magnetic_Flux", unit: "Wb",desc: "magnetic flux", use: "G"},
        "046": {name: "DPT_Value_Magnetic_FluxDensity", unit: "T",desc: "magnetic flux density", use: "G"},
        "047": {name: "DPT_Value_Magnetic_Moment", unit: "A m2",desc: "magnetic moment", use: "G"},
        "048": {name: "DPT_Value_Magnetic_Polarization", unit: "T",desc: "magnetic polarization", use: "G"},
        "049": {name: "DPT_Value_Magnetization", unit: "A m-1",desc: "magnetization", use: "G"},
        "050": {name: "DPT_Value_MagnetomotiveForce", unit: "A",desc: "magneto motive force", use: "G"},
        "051": {name: "DPT_Value_Mass", unit: "kg",desc: "mass", use: "G"},
        "052": {name: "DPT_Value_MassFlux", unit: "kg s-1",desc: "mass flux", use: "G"},
        "053": {name: "DPT_Value_Momentum", unit: "N s-1",desc: "momentum", use: "G"},
        "054": {name: "DPT_Value_Phase_AngleRad", unit: "rad",desc: "phase angle, radiant", use: "G"},
        "055": {name: "DPT_Value_Phase_AngleDeg", unit: "° 1° phase angle, degrees", use: "G"},
        "056": {name: "DPT_Value_Power", unit: "10) W",desc: "power", use: "G"},
        "057": {name: "DPT_Value_Power_Factor", unit: "cos Φ",desc: "Φ power factor", use: "G"},
        "058": {name: "DPT_Value_Pressure", unit: "Pa = N m-2",desc: "pressure", use: "G"},
        "059": {name: "DPT_Value_Reactance", unit: "Ω", desc: "reactance", use: "G"},
        "060": {name: "DPT_Value_Resistance", unit: "Ω", desc: "resistance", use: "G"},
        "061": {name: "DPT_Value_Resistivity", unit: "Ωm", desc: "resistivity", use: "G"},
        "062": {name: "DPT_Value_SelfInductance", unit: "H",desc: "self inductance", use: "G"},
        "063": {name: "DPT_Value_SolidAngle", unit: "sr",desc: "solid angle", use: "G"},
        "064": {name: "DPT_Value_Sound_Intensity", unit: "W m-2",desc: "sound intensity", use: "G"},
        "065": {name: "DPT_Value_Speed", unit: "m s-1",desc: "speed", use: "G"},
        "066": {name: "DPT_Value_Stress", unit: "Pa = N m-2",desc: "stress", use: "G"},
        "067": {name: "DPT_Value_Surface_Tension", unit: "Nm-1 1 Nm-1 surface tension", use: "G"},
        "068": {name: "DPT_Value_Common_Temperature", unit: "°C 1°C temperature, common", use: "G"},
        "069": {name: "DPT_Value_Absolute_Temperature", unit: "K vK temperature (absolute)", use: "G"},
        "070": {name: "DPT_Value_TemperatureDifference", unit: "K",desc: "temperature difference", use: "G"},
        "071": {name: "DPT_Value_Thermal_Capacity", unit: "JK-1",desc: "thermal capacity", use: "G"},
        "072": {name: "DPT_Value_Thermal_Conductivity", unit: "W m-1  K-1",desc: "thermal conductivity", use: "G"},
        "073": {name: "DPT_Value_ThermoelectricPower", unit: "V K-1",desc: "thermoelectric power", use: "G"},
        "074": {name: "DPT_Value_Time", unit: "s",desc: "time 11)", use: "G"},
        "075": {name: "DPT_Value_Torque", unit: "Nm",desc: "torque", use: "G"},
        "076": {name: "DPT_Value_Volume", unit: "m3",desc: "volume", use: "G"},
        "077": {name: "DPT_Value_Volume_Flux", unit: "m3 s-1",desc: "volume flux", use: "G"},
        "078": {name: "DPT_Value_Weight", unit: "N",desc: "weight", use: "G"},
        "079": {name: "DPT_Value_Work", unit: "J",desc: "work", use: "G"}
    }
};
